autofit/graphical/messages/abstract.py

```diff
@@ -1,256 +1,389 @@
+        elif shape[1:] == self.shape:
+            eta = self.natural_parameters
+            t = self.to_canonical_form(x)
+            eta_t = np.multiply(eta[:, None, ...], t).sum(0)
+            return self.log_base_measure + eta_t - self.log_partition
+            # return np.array([self.logpdf(x_) for x_ in x])
+
+        raise ValueError(
+            f"shape of passed value {shape} does not match message shape {self.shape}")
+
+    def numerical_logpdf_gradient(self, x: np.ndarray, eps: float=1e-6
+    ) -> Tuple[np.ndarray, np.ndarray]:
+        shape = np.shape(x)
+        if shape:
+            x0 = np.array(x, dtype=np.float64)
+            logl0 = self.logpdf(x0)
+            if self._multivariate:
+                grad_logl = np.empty(logl0.shape + x0.shape)
+                sl = tuple(slice(None) for _ in range(logl0.ndim))
+                with np.nditer(x0, flags=['multi_index'], op_flags=['readwrite']) as it:
+                    for xv in it:
+                        xv += eps
+                        logl = self.logpdf(x0)
+                        grad_logl[sl + it.multi_index] = (logl - logl0)/eps
+                        xv -= eps
+            else:
+                l0 = logl0.sum()
+                grad_logl = np.empty_like(x0)
+                with np.nditer(x0, flags=['multi_index'], op_flags=['readwrite']) as it:
+                    for xv in it:
+                        xv += eps
+                        logl = self.logpdf(x0).sum()
+                        grad_logl[it.multi_index] = (logl - l0)/eps
+                        xv -= eps
+        else:
+            logl0 = self.logpdf(x)
+            grad_logl = (self.logpdf(x + eps) - logl0)/eps
+
+        return logl0, grad_logl
+
+    def numerical_logpdf_gradient_hessian(self, x: np.ndarray, eps: float=1e-6
+    ) -> Tuple[np.ndarray, np.ndarray]:
+        shape = np.shape(x)
+        if shape:
+            x0 = np.array(x, dtype=np.float64)
+            if self._multivariate:
+                logl0, gradl0 = self.numerical_logpdf_gradient(x0)
+                hess_logl = np.empty(gradl0.shape + x0.shape)
+                sl = tuple(slice(None) for _ in range(gradl0.ndim))
+                with np.nditer(x0, flags=['multi_index'], op_flags=['readwrite']) as it:
+                    for xv in it:
+                        xv += eps
+                        _, gradl = self.numerical_logpdf_gradient(x0)
+                        hess_logl[sl + it.multi_index] = (gradl - gradl0)/eps
+                        xv -= eps
+            else:
+                logl0 = self.logpdf(x0)
+                l0 = logl0.sum()
+                grad_logl = np.empty_like(x0)
+                hess_logl = np.empty_like(x0)
+                with np.nditer(x0, flags=['multi_index'], op_flags=['readwrite']) as it:
+                    for xv in it:
+                        xv += eps
+                        l1 = self.logpdf(x0).sum()
+                        xv -= 2 * eps
```

Comment:
Looks like l1 == l2?

Reply:
`xv` is a single element of the array `x0` so it's been changed by +/-  `eps`